### DISCUSSION TITLE
should we provide a local, high-res topo dataset?

### DISCUSSION THREAD

**Discussion by dankelley** [dankelley/oce#1756] [Q&A] [answered]
I was musing (https://github.com/dankelley/oce/pull/1755#issuecomment-743178604) that it might be nice to have a higher-resolution topography dataset.

The problem with that is that even reducing the bin width by a factor of 2 increases the file size by a factor of 4, which is a problem because this is the biggest dataset in oce already.

Another idea might be to provide a high-resolution file for a small region (I would say, the Maritimes), so that folks could see the sort of resolution that is easily downloadable nowadays.

I'm sort of thinking out loud here.  This is my first 'discussion' item, and partly I am posting it here and not as an issue because I view 'issues' as 'problems', and my goal here is, literally, to promote discussion.

**Comment by richardsc** [comment]
I like this idea. I'm pretty used to calling up marmap to get what I want, but it requires a connection (and leaves cast-off files in its wake) and adds a few lines to what should be a simple example. The "Northwest Corner" would definitely be a good candidate.

**Reply by dankelley** [reply]
By 'northwest corner' do you mean Flemish cap, as in Lazier (1994)?  I was sort of thinking also of showing near Halifax, as advertisement ... although the oceanography is much more interesting near Flemish Cap so, frankly, I prefer your suggestion.

**References** Lazier, John R. N. “Observations in the Northwest Corner of the North Atlantic Current.” Journal of Physical Oceanography 24, no. 7 (July 1, 1994): 1449–63. https://doi.org/10.1175/1520-0485(1994)024<1449:OITNCO>2.0.CO;2.

**Reply by richardsc** [reply]
I guess I wasn't thinking that specific -- more like as much of the Northwestern Atlantic (including Grand Banks, Scotian Shelf, SLE, etc) as we think we could fit in a high rest set. Perhaps at highest resolution we'd need to keep it limited though.

**Reply by dankelley** [reply]
I fiddled a bit, coming up with as follows.  The plot title gives the size, which I think is okay, and the res is not too 'patchy', I think.  (My view is that adding another file under 200K won't be a problem but if we double the resolution and thus make the file be 4X bigger, then we'd likely want to lower the domain span.)

I wonder if other oce developers think that this region is interesting, and if the resolution seems okay. Any thoughts, @richardsc and @clayton33?  We want something that will give a graph that people will find interesting to look at, and I like the idea of showing local shelf waters, even if (of course) the shelves are nearly monochromatic, owing to the scale being controlled by a desire to capture variation in the real ocean.

We could also go for much finer resolution over a smaller domain.  I think Sable Island and 'The Gulley' would be neat to see, and of interest. (All oceanographers should know of Sable Island.) Or any other spot that you folks think would be neat.  The easiest would be to start with the code I give below.  (Hint: reprex makes it really easy to show code and a figure...)


``` r
## Test code for possible local high-res topo dataset
library(oce)
#> Loading required package: gsw
#> Loading required package: testthat
#> Loading required package: sf
#> Linking to GEOS 3.8.1, GDAL 3.2.0, PROJ 7.2.0
data(coastlineWorldFine, package="ocedata")
topoFile <- download.topo(-67,-42,40,54, resolution=8) # nwatl
topo <- read.topo(topoFile)
cm <- colormap(c(-5000,0), col=oceColorsGebco)
imagep(topo, colormap=cm, xlab="", ylab="", mar=c(2,2,1,1))
polygon(coastlineWorldFine[["longitude"]], coastlineWorldFine[["latitude"]], col="lightgrey")
mtext(system(paste("ls -lh", topoFile), intern=TRUE), cex=0.7)
```

![](https://i.imgur.com/7RBeSAN.png)

<sup>Created on 2020-12-11 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

**Reply by dankelley** [reply]
Here's another trial.  (I cannot use `reprex` because the image won't upload to imgur ... maybe it's too big, or there is a limit per day, or imgur is down.)

In this one, I'm not getting into deep ocean, so you can see more detail on the shelves.  Also I'm showing the land because it matters for coastal work.

NOTE: something is a bit off with the `oceColorsGebco` colourmap (I'll make an issue), because very shallow water is coming out with a landish colour.

```R
## Test code for possible local high-res topo dataset
library(oce)
data(coastlineWorldFine, package="ocedata")
topoFile <- download.topo(-67,-59.5,43.3,47.2, resolution=2) # NS
## topoFile <- download.topo(-67,-42,40,54, resolution=8) # nwatl
topo <- read.topo(topoFile)
H <- quantile(abs(topo[["z"]]), 0.95)
cm <- colormap(H*c(1, -1), col=function(n) oceColorsGebco(region="both",n=n))
imagep(topo, colormap=cm, xlab="", ylab="", mar=c(2,2,1,1))
lines(coastlineWorldFine[["longitude"]], coastlineWorldFine[["latitude"]])
#polygon(coastlineWorldFine[["longitude"]], coastlineWorldFine[["latitude"]], col="lightgrey")
mtext(system(paste("ls -lh", topoFile), intern=TRUE), cex=0.7)
```

![image](https://user-images.githubusercontent.com/99469/101932420-1d320680-3bb1-11eb-9d5d-f965e22f4931.png)

**Reply by richardsc** [reply]
I like it! Interested to see what the color issue is ... have you tried with `colormap()`?

**Reply by dankelley** [reply]
Yup, I'm using `colormap()`.  The problem is in how I define the gebco transitions.  I'll fiddle with it today, since I have a break from grading.

**Reply by dankelley** [reply]
Note to self -- the image below shows the colours that get interpolated between for the gebco scheme.  Neither land nor water contains pure white, so it may not be surprising that interpolating in colourspace might give funny answers.  But there are other subtle things, e.g. do we count from 0 or from 1, and in which colourspace should interpolation be done.  Again, this is just a note to self, so I can see it later if needed.

![image](https://user-images.githubusercontent.com/99469/102085123-96696d80-3dec-11eb-8d07-ae5d879585cf.png)

**Reply by dankelley** [reply]
(Question for @clayton33, @richardsc and @j-harbin below ... no real need to read the whole thread, unless you're interested. My question relates to the plot that you'll see above this, if you look on GH. Thanks!!)

I am playing with adding one or two white bands between the land and sea bands.  Again, none of this is on GH -- just notes to self.

I think I like one white band, but I also want to add a lighter-tan colour for low-elevation land, because to my eye, the lightest of the blues is lighter than the lightest of the browns.

**QUESTION for co-devels:** am I right that the lightest blue is ligher than the lightest tan?  I am not sure I see colours correctly. In particular, I am interested in the opinion of @clayton33, who seems to see colours better than I can.  I'd also like to hear from @richardsc and @j-harbin, who have good eyes for aesthetics.

*OLD SCHEME*

![00topoA](https://user-images.githubusercontent.com/99469/102088083-cb77bf00-3df0-11eb-9db5-69e902e648bf.png)


*WITH ONE WHITE BAND ADDED*

![00topoB](https://user-images.githubusercontent.com/99469/102088229-04b02f00-3df1-11eb-9a44-df3d4b4fd793.png)


*WITH TWO WHITE BANDS ADDED*

![00topoB](https://user-images.githubusercontent.com/99469/102088120-dc283500-3df0-11eb-9dd1-8da372c2da88.png)

**Reply by clayton33** [reply]
You're right, the lightest blue is lighter than the lightest tan (confirmed quickly and kind of dirty by taking a photo of my computer screen with my phone and then changing the photo to black and white).

HmnNnn... the two white bands makes the colour map more 'centered' around 0, if that is what you're going for. Only adding 1 still makes it look lopsided. But maybe a few test map plots will tell the truth, perhaps adding one white band will be enough remove those land colours for shallow waters. Two might bleed too much onto land, then you'll kind of be in the same place that you started ?

**Reply by dankelley** [reply]
@clayton33 What do you think of the attached, in which the upper row is my new idea for the colours? I lightened the land on the light end, and also added yet-whiter land and water.  The idea of the addition is that it will make the band of white less thick, if I add a single colour there (and, yes, I agree that a single will be better than a double).

If the colours look ok, I'll take the next step and start to code it in, so I can see a palette and the NS image.  But the first step is to get colours that seem balanced.

I appreciate the help!

![01](https://user-images.githubusercontent.com/99469/102097028-c66c3d00-3dfb-11eb-808e-7cec33d0c870.png)

**Reply by clayton33** [reply]
Ooo, yes, I like that proposed scheme. Sorry it took an hour to respond, was deep into some coding and didn't check e-mail.

**Reply by richardsc** [reply]
We're kind of way off topic here (the discussion was supposed to be about a new dataset, not about the Gebco colors 😄 ), but I just pointed out to Dan in an email that there is also the existing `gmt_relief` colormap, as well as `topo` in the `cmocean` package.

One thing you'll notice about both of those is they have a "sharp" transition at the water/land boundary, to make it clear that it's dry vs wet:

![image](https://user-images.githubusercontent.com/233584/102104687-dd635d00-3e04-11eb-9905-586aea034c95.png)

**Reply by dankelley** [reply]
I'm getting somewhere.  I will return to the actual application in a while, but at the moment I have three contenders, as shown below. The one on the left is the existing algorithm, but with a white band added in the middle.  I think the white region is too thick.  The other two are slight variants of a scheme without a white band, but with lightened brown.  (The difference is in the colorspace within which interpolation is done.)

I am leaning towards the middle one, or the right one.  I *think* I see very slight brownishness to the lightest of the blue, though, and so I may lighten the lightest brown just a tad.

![topo_colors_02](https://user-images.githubusercontent.com/99469/102121519-781a6680-3e1a-11eb-9e32-0a51ef82c37f.png)

**Reply by richardsc** [reply]
Have you tried plotting a map with it though? Does it look good around the coastline elevations?

What if you do the blue-to-white for the water, and then do darkbrown-to-lightbrown for land (so that the lowest elevations have the darkest brown?

**Reply by dankelley** [reply]
We have lowest land as darkest brown with the other schemes.  The gebco scheme (well, the one mimicked by oce for the past years or maybe decade) uses light brown in low altitudes.  I just checked the gebco graphics at https://www.gebco.net/data_and_products/printable_maps/documents/gebco_world_map_2014.pdf (a large download) and it's interesting, because it looks as though they are not actually coding land for elevation; it seems they are coding with a shading model, and possibly colourizing green or brown according to whether there is vegetation.  I've put a zoomup of the NS region below.  Note that Cape Breton has elevations of approx 400m, but the colour doesn't look different from colours in PEI, which has much lower elevations.

If gebco published official colours coded to elevation, we should use them.  But I cannot find them.

Since we have the cmocean and gmt_ colours available already, what I am doing is more in the line of debugging some code, and improving some aesthetics.  I just don't like an ugly plot, and colourizing water with a landish hue bothers me :-)

<img width="574" alt="Screen Shot 2020-12-14 at 3 18 17 PM" src="https://user-images.githubusercontent.com/99469/102125193-a9496580-3e1f-11eb-89c2-ec8bf82b2751.png">

**Reply by richardsc** [reply]
Yes, I went looking for a published Gebco scheme, but found like you did. So, carry on! 😄

**Reply by dankelley** [reply]
The problem is basic colour theory.  Blue cannot cancel red, so when we mix a very light blue with a very light reddish brown, the result has a slight reddish hue.  Since we devise colours by interpolating in colour space, we get a problem.  I'm fiddling with ways around that.  One is to put some white paint on the palette, but to make that work without a pretty thick white region near the coasts, I will want to add more colours on the sides.  Right now I have 9 colours on each side.  I think I got that by digitizing a gebco chart that had a colourscale.  I am not going to beg the department chair for his approval to beg the Dean to go into work go get a chart, to start a new scan.  @richardsc if you happen to be at work and have a chart and a scanner, maybe you could scan the colour-bar portion ... :-)

**Reply by dankelley** [reply]
@clayton33 -- how does the following look?  The Q is whether there is a brownish hue in waters south of Shelburne (southern tip of NS).  I don't see that hue there anymore, but I have troubles with some colours, it seems! I'm not worried about a few pixels that don't line up with the coastline (like in Cape Breton); these are inevitable near steep topography unless you zoom in the topo a lot.

This uses 1 minute topo (pixel size 1.8km in north-south direction, and 1.4km in east-west direction, at this lat). The .rda file is 159K, so I think that would be okay to include this.

I think this is a lovely plot, actually.  Part of it is that I know the land reasonably well, having driven up the various hills (mountains) and along low-lying areas, too.  The familiarity makes it fun to look at.  And I want to advertise the Maritimes in this dataset.

I find the detail fascinating, e.g.
* There are some blue smudges in the middle of NB, because the topo shows the bottom of the lakes, not the top, and those lakes are at sea level.  (See also https://en-ca.topographic-map.com/maps/qun/New-Brunswick/ ... which has a neat colour scale for land ... oh no, another rabbithole)
* You can see the LaHave river valley going pretty far inland.
* Cape Breton has some high land.  I can see why folks from there look down on the rest of us.
* Speaking of being looked-down upon, the pic shows the low land of ~God's country~ the Tantramar Marshes clearly.

PS. the code for this is not on GH yet.  I would like to get some more eyes on this first (and I apologize for taking up folks' time).

![topo_NS](https://user-images.githubusercontent.com/99469/102130030-aaca5c00-3e26-11eb-9a9a-48494114f0c0.png)

**Reply by clayton33** [reply]
Oooo, this looks nice.

**Reply by richardsc** [reply]
Ok yes, that does look pretty good.

I'd like to point out that probably much of this "discussion" should probably have been done inside issue #1757 .... just saying 😄

**Reply by dankelley** [reply]
The funny thing is that I made that issue, but then wrote comments here.  Oh well.  This was a test of GH discussions.  I like the way it hides stuff in the 'read more' link.  That's a form of threading, I guess.  But this is not the kind of threading I was really hoping for in GH discussions.  I was hoping that e.g. there would be a little triangle symbol for this whole sidetracking.  Many issues end up with sidetracking.

Anyway, the new datafile is only 87K so I think that should be fine for adding.

Unless someone objects, I'm going to go ahead and insert that into oce this today at 1300h.  I'll write up some docs for it, of course.  And I'll see whether some examples can use it, instead of `topoWorld`.  If I were a user, and saw a pretty map like this, I'd be excited to try it myself.

PS. I use `download.topo()`, which gets a netcdf file that is much smaller than the csv that `marmap` downloads, and that loads faster, too, because it's already in matrix form.  But the server changed, and so the `download.topo()` from the CRAN version of oce won't work. Yet another reason to consider a new CRAN release in January, I think.

Oh, and the vignette could use it, too. For example, it might be neat to write a bit of code to go along the Halifax line and extract bathymetry along it.   Possibly doing such an extraction might be a useful exercise for @j-harbin, who knows people who know the lon-lat of the Halifax line...

**Reply by dankelley** [reply]
I've added `data(topoNS)`, and `example(topoNS)` now plots it, first in contour form and then in image form.

**Comment by dankelley** [comment, marked as answer]
I gave up on this idea because, during the last CRAN release, I wanted to remove every unneeded byte from the package, to get in under the size limit.
